ci/test.sh

```diff
@@ -1,101 +1,159 @@
-# a hex number. The Mac version is newer so it has a different output format.
-if [ "$TRAVIS_OS_NAME" = "osx" ]; then
-	export GITTEST_REMOTE_SSH_FINGERPRINT=$(ssh-keygen -E md5 -F '[localhost]:2222' -l | tail -n 1 | cut -d ' ' -f 3 | cut -d : -f2- | tr -d :)
-else
-	export GITTEST_REMOTE_SSH_FINGERPRINT=$(ssh-keygen -F '[localhost]:2222' -l | tail -n 1 | cut -d ' ' -f 2 | tr -d ':')
+	EOF
+	ssh-keygen -t rsa -f "${SSHD_DIR}/id_rsa" -N "" -q
+	/usr/sbin/sshd -f "${SSHD_DIR}/sshd_config" -E "${SSHD_DIR}/log"
+
+	# Set up keys
+	mkdir "${HOME}/.ssh"
+	ssh-keygen -t rsa -f "${HOME}/.ssh/id_rsa" -N "" -q
+	cat "${HOME}/.ssh/id_rsa.pub" >>"${HOME}/.ssh/authorized_keys"
+	while read algorithm key comment; do
+		echo "[localhost]:2222 $algorithm $key" >>"${HOME}/.ssh/known_hosts"
+	done <"${SSHD_DIR}/id_rsa.pub"
+
+	# Get the fingerprint for localhost and remove the colons so we can
+	# parse it as a hex number. Older versions have a different output
+	# format.
+	if [[ $(ssh -V 2>&1) == OpenSSH_6* ]]; then
+		SSH_FINGERPRINT=$(ssh-keygen -F '[localhost]:2222' -f "${HOME}/.ssh/known_hosts" -l | tail -n 1 | cut -d ' ' -f 2 | tr -d ':')
+	else
+		SSH_FINGERPRINT=$(ssh-keygen -E md5 -F '[localhost]:2222' -f "${HOME}/.ssh/known_hosts" -l | tail -n 1 | cut -d ' ' -f 3 | cut -d : -f2- | tr -d :)
+	fi
 fi
 
 # Run the tests that do not require network connectivity.
 
-echo ""
-echo "################################################################################"
-echo "## Running (non-online) tests                                                 ##"
-echo "################################################################################"
+if [ -z "$SKIP_OFFLINE_TESTS" ]; then
+	echo ""
+	echo "##############################################################################"
+	echo "## Running (offline) tests"
+	echo "##############################################################################"
 
-#ctest -V -R offline
+	ctest -V -R offline || die $?
+fi
 
-# Run the various online tests.  The "online" test suite only includes the
-# default online tests that do not require additional configuration.  The
-# "proxy" and "ssh" test suites require further setup.
+if [ -z "$SKIP_ONLINE_TESTS" ]; then
+	# Run the various online tests.  The "online" test suite only includes the
+	# default online tests that do not require additional configuration.  The
+	# "proxy" and "ssh" test suites require further setup.
 
-echo ""
-echo "################################################################################"
-echo "## Running (online) tests                                                     ##"
-echo "################################################################################"
+	echo ""
+	echo "##############################################################################"
+	echo "## Running (online) tests"
+	echo "##############################################################################"
 
-#ctest -V -R online
+	ctest -V -R online || die $?
```

Comment:
Do we really want to die if any of those tests fails? I think it would be more sensible to just record that a test has failed and die after _all_ tests ran. If you die eagerly, you'll potentially have to re-run CI every time you have fixed a test just to stumble into the next one that was previously not executed.

Reply:
This is what we're doing today.  🤷‍♂️ I would rather have us continue to fail fast, on the hopes that we get out of the CI queue as quickly as possible.  I guess that could lead to someone spamming the queue with brokenness, but we haven't seen that so far.